kernel/src/main/java/org/kframework/parser/inner/ParseInModule.java

```diff
@@ -300,18 +300,15 @@ public Scanner getScanner() {
             return new Tuple2<>(Right.apply(rez3), warn);
         } finally {
             if (profileRules) {
-                long stop = System.nanoTime();
                 try {
                     Writer t = timing;
                     synchronized(t) {
-                        t.write(source.toString());
-                        t.write(':');
-                        t.write(Integer.toString(startLine));
-                        t.write(':');
-                        t.write(Integer.toString(startColumn));
-                        t.write(' ');
-                        t.write(Double.toString((stop - start) / 1000000000.0));
-                        t.write('\n');
+                        long stop = System.currentTimeMillis();
```

Comment:
By moving the stop time measurement inside the lock, you have added the totally irrelevant time spent waiting on the lock to the measurement.

Reply:
That's fair. BUT, that's also confusing. On #2116 you argued that using a `ConcurrentLinkedDeque` which is thread-safe but contentious when adding elements is not ok. But you are using a lock here.

What is going on? How is this code working/supposed to work?

Reply:
You're right that something weird is going on. When I first wrote this code, each thread had its own ParseInModule instance so it didn't really matter that there was a lock there. At some point that changed so each thread uses the same ParseInModule instance, and we didn't address this inconsistency.  You probably want to change the writer so it's inside a ThreadLocal, I think, and remove the lock.

Reply:
I see 4 possible solutions here:
1. Use ThreadLocal to print the timing info in a file per thread. First google link I found though warns about using ThreadLocal+ThreadPools: https://www.baeldung.com/java-threadlocal#threadlocalsand-thread-pools
2. Generate a new tempfile for each parsed rule. Basically, bypass the synchronization problem entirely. We're combining them into a single file anyway. But we create more IO traffic.
3. Use a `ConcurrentLinkedQueue` which (as I said above) is threadsafe but contentious when adding elements. But the implementation is based on volatility ([Compare-And-Swap](https://www.baeldung.com/java-queue-linkedblocking-concurrentlinked#concurrentlinkedqueue)). Meaning the variable it's writing into is _not_ stored in cache, only in the main memory, and it offers very low overhead and low probability of collision. Since we're doing this operation outside of the area we're measuring, it's fine. I see this solution as a lower cost than writing to disk.
4. `Tuple2<Errors, Term> parseInModule` becomes `Tuple3<Errors, Term, Timing> parseInModule`. In other words, bypass the thread synchronization issue, and let the stream collectors put everything together. I may even be able to add timing info for rules that fail to parse.

Reply:
I like the second solution. We are only appending data so the I/O overhead should be pretty low since it will be sequential on disk. We can explore a different solution if it seems that the data changes too much based on this.